### ISSUE TITLE
Support for Firefox OS

### ISSUE THREAD

**Issue by airon90** [syncthing/syncthing#705]
Hi, do you want to support Firefox OS by creating an app?

Here you have all info to create an app for FxOS and how to obtain a free smartphone to code for it:
- https://hacks.mozilla.org/2014/05/build-your-next-app-with-a-flame/
- https://marketplace.firefox.com/developers/

**Comment by Eucyon** [comment]
You can temporarily install and use Syncthing (ST) on Firefox Os (FFOS) devices as follows. Temporarily means that by the next booting, the installation is deleted.

This small guide is made for Ubuntu-Linux. Where necessary, adapt it to your operating system. Follow the guide on your own risk.

Prerequisite is ADB on your computer.

Perform the following steps:

1) Download the latest ST version for your device under https://github.com/syncthing/syncthing/releases. Most FFOS devices have ARMv7 CPUs.

2) Unpack the downloaded file.

3) Enable ADB debugging in the developer settings of the device and connect it to your computer.

4) Push the unpacked folder to the device. Therefore open a terminal on your computer and type:
adb push (Path/to/unpacked_file) /data/local/tmp/syncthing

5) Start ST on the device in your computer's terminal with:
adb shell (Enter)
/data/local/tmp/syncthing/syncthing -home="/data/local/tmp/syncthing" & (Enter)

ST is now running on the phone. Open a browser there and reach ST's Webgui via the address 127.0.0.1:8080. You can copy the keys for the settings to make on your computer from the output of step 5. On the device they are proposed.

This worked for me on a ZTE Open C with FFOS 1.3 months ago and now on an Alcatel Fire E with FFOS 2.0. But:

6) Meanwhile issue #831 occured, which also applies to FFOS devices. I could fix it by changing the config.xml file in the ST folder. Stop ST in the webgui of the device. On your computer, pull the file to your computer with:

adb pull /data/local/tmp/syncthing/config.xml config.xml

On your computer, open the file with an editor an change

        <lenientMtimes>false</lenientMtimes>

to

        <lenientMtimes>true</lenientMtimes>

for the folders to be synced, save it and push the config.xml back to the device:

adb push config.xml /data/local/tmp/syncthing/config.xml. Start ST again, as described in step 5).

At the moment, from my understanding, this is the only way to run ST on FFOS, because its apps are written in Java and the Java layer can't control this Linux level.

**Comment by kozec** [comment]
@Eucyon Does ST launched like that has access to user files? If yes, it should be possible to write Firefox OS wrapper similar to Syncthing-Android.

**Comment by Eucyon** [comment]
Dear kozek,

I am syncing my sd-card (/storage/sdcard) to my Ubuntu computer. In this sense ST has access to user files.

But a wrapper would have to be programmed in Java, and the Java layer has no direct control over the Linux level. As long as there are no FFOS-APIs for such things, a wrapper could only be realised with serious hacking, e.g. a custom ROM. And this is beyond my expertise at the moment.

**Comment by acidicX** [comment]
@Eucyon I think you mean JavaScript (huge difference!)?
I've created a gist where I automated the process a bit. It also does not use the tmp dir, therefore it is still there after a reboot. You still need to trigger the shell command after a reboot though, because I haven't found any init-files on b2g yet.
https://gist.github.com/acidicX/17b826310530310c17d8

@kozec everything launched by adb shell is root, therefore access to files is no problem - but I would not recommend running ST as root... AFAIK there is no way around that though. I'm not sure if it is wise to try to run it as the b2g (UI) user. There is a user called media_rw, which owns the sdcard, maybe I can try that...

Still, the ultimate solution is to compile to JavaScript for a native FxOS app.

**Comment by Eucyon** [comment]
@acidicX
Javascript, of course!

Your improvement works fine on rooted phones. Mine is not yet, I can't remount.

In my eyes it is quite extensive to port ST to Javascript, in any case for me. On the other hand, stock FFOS doesn't offer a clean way to start binaries on the phone from the JS-level. So your idea (which I also had, but haven't detailed by now) with the init looks good to me.

FFOS does have an init.rc: https://developer.mozilla.org/en-US/Firefox_OS/Platform/Architecture

On my Fire E it is located at /init.rc.

**Comment by acidicX** [comment]
@Eucyon I only got a Flame. That comes with root :-)
Right, I've just tried using the init.rc and init.b2g.rc but it won't load after reboot. No errors though...
Anyway, you'll just never get a Go binary packaged into FxOS. All gaia components (apps) need to be JS/HTML/CSS, for a reason (permissions, etc.). The only real solution is to compile to JS. As the frontend of ST is already implemented as a web service, I guess it's easier (than Android) since you don't even need a GUI wrapper. As for the performance, you are not likely to sync extreme amounts of data with a phone. This is pure speculation, but I think I could live with the performance impact.

If anybody with experience in Go could take a look at GopherJS, that would be great (create a list that needs to be done to compile...). Maybe I can do some of the work (filesystem, as mentioned). At the moment, I would not know where to start, but I can volunteer for testing, as I've got a Flame dev device for FxOS.

@calmh maybe add the "help needed" tag instead of wontfix?
